src/decoder/mod.rs

```diff
@@ -723,27 +723,28 @@ impl<R: Read> Reader<R> {
     /// The scanline is filtered against the previous scanline according to the specification.
     fn next_raw_interlaced_row(&mut self, rowlen: usize) -> Result<(), DecodingError> {
         // Read image data until we have at least one full row (but possibly more than one).
-        while self.current.len() - self.scan_start < rowlen {
+        while self.data_stream.len() - self.current_start < rowlen {
             if self.subframe.consumed_and_flushed {
                 return Err(DecodingError::Format(
                     FormatErrorInner::NoMoreImageData.into(),
                 ));
             }
 
             // Clear the current buffer before appending more data.
-            if self.scan_start > 0 {
-                self.current.drain(..self.scan_start).for_each(drop);
-                self.scan_start = 0;
+            if self.prev_start > 0 {
+                self.data_stream.drain(..self.prev_start).for_each(drop);
```

Comment:
The `drain` / `for_each(drop)` approach seems complex - I wonder if the compiler is able to optimize this as well as a more direct move - e.g. something like:

```
            self.data_stream.copy_within(self.prev_start.., 0);
            self.data_stream.truncate(self.data_stream.len() - self.current_start);
```

I can't really read assembly well, but the `drain` approach seems to at least result in longer assembly: https://godbolt.org/z/z48q1xej5

OTOH, maybe this feedback doesn't belong to this PR and we should fix this in a separate PR?  (e.g. `drain`-based compaction is also used in `zlib.rs`)